Create eventing mesh overview and architecture (#7278)
kyma-project/kyma: docs/knative-eventing-mesh/02-01-knative-eventing-mesh.md

---
title: Architecture
---

The architecture of Knative Eventing Mesh relies heavily on the functionality provided by [Knative Eventing](https://knative.dev/docs/eventing/). To ensure a stable event flow between the Sender and the Subscriber, the Knative Eventing Mesh wires the Knative elements closely with the existing Kyma components.

## Knative Eventing Mesh implementation

This diagram shows how the Knative elements are implemented and wired with the already exisitng Kyma components, whereas the second diagram shows you how the events are processed by in Kyma.

![Eventing implementation](./assets/eventing-mesh-implementation.svg)

> 1. The user creates the [Application CR](https://kyma-project.io/docs/components/application-connector/#custom-resource-application) and binds it to the Namespace. 

Reviewer: Please add blank line after every point and every subpoint  :) This is a good practice.

Author: 👍

2. The Application Operator watches the creation of an Application CR and creates the HTTP Source, which defines the event sources the events can come from. 
3. The Event Source Controller watches the creation of the HTTP Source and deploys these resources:
    * [HTTP source adapter](https://github.com/kyma-project/kyma/tree/master/components/event-sources/adapter/http) which is an HTTP server deployed inside the `kyma-integration` Namespace. The adapter acts as a gateway to the Knative Channel, and its responsibility is to expose an endpoint the Application sends the events to. 
    * [Knative Channel](https://knative.dev/docs/eventing/channels/) CR which defines persistence layer and is responsible for dispatching events to multiple destinations. 
4. The Application Broker watches the creation of the Application CR and performs the following actions:

>     * Exposes event definitions as an event ServiceClass. Once the Application Broker provisions the ServiceClass, the events become available for services. 

Reviewer: Please add blank line after every subpoint  :) This is a good practice.

    * Creates the Knative Subscription CR which enables forwarding events received by the Knative Channel to Subscribers or other Channels. 
    * Injects the [Knative Broker](https://knative.dev/docs/eventing/broker-trigger/) to the User Namespace. The Broker acts as an entry point for the events. 
5. The user creates the [Knative Trigger](https://knative.dev/docs/eventing/broker-trigger/) which defines the subscriber along with the conditions for filtering out events. This way, certain subscribers receive only the events they are interested in. For details on Trigger CR specification, see the **Trigger Filtering** section of [this](https://knative.dev/docs/eventing/broker-trigger/) document.

## Event flow 

This diagram explains the event flow in Kyma, from the moment the Application sends an event to the point where it triggers a specific function.

![Eventing flow](./assets/eventing-mesh-flow.svg)

> 1. The Application sends events to endpoint speficed in the HTTP source adapter definition.

Reviewer: Please add blank line after every point  :) This is a good practice.

---
title: Architecture
---

The architecture of Knative Eventing Mesh relies heavily on the functionality provided by [Knative Eventing](https://knative.dev/docs/eventing/). To ensure stable event flow between the sender and the subscriber, Knative Eventing Mesh wires Knative elements closely with the existing Kyma components.

> ## Knative Eventing Mesh implementation

Reviewer 2: or `## Knative Eventing Mesh components`

> ## Knative Eventing Mesh implementation

Reviewer 2: I rather meant `## Component dependencies`

1. The user creates the [Application CR](https://kyma-project.io/docs/components/application-connector/#custom-resource-application) and binds it to the Namespace. 

2. The Application Operator watches the creation of the Application CR and creates the [HTTPSource CR](#custom-resource-http-source) which defines the source sending the events.

3. The Event Source Controller watches the creation of the HTTPSource CR and deploys these resources:

    * [HTTP Adapter](https://github.com/kyma-project/kyma/tree/master/components/event-sources/adapter/http) which is an HTTP server deployed inside the `kyma-integration` Namespace. The adapter acts as a gateway to the Knative Channel, and is responsible for exposing an endpoint the Application sends the events to. 

    * [Knative Channel](https://knative.dev/docs/eventing/channels/) which defines the way messages are dispatched in the Namespace. Its underlying implementation, such as NATS Streaming or Kafka Channel, is responsible for forwarding events to multiple destinations. 

4. The Application Broker watches the creation of the Application CR and performs the following actions:

>     * Exposes event definitions as an event ServiceClass. Once the user deploys the Service using this ServiceClass, the Application Broker provisions it to make events available for services.

Reviewer 2: I would still specify if there is one class per one destination?

## Event flow 

This diagram explains the event flow in Kyma, from the moment the Application sends an event, to the point when the event triggers the function.

![Eventing flow](./assets/eventing-mesh-flow.svg)

1. The Application sends events to the HTTP Adapter which forwards them to the resource such as Knative Broker.
   
    >**NOTE:** The HTTP adapter accepts only CloudEvents in version 1.0. 

2. Knative Subscription defines the Broker as the Subscriber. This way, Knative Channel can communicate with the Broker to send events.

3. Knative Channel listens for incoming events. When it receives an event, the underlying messaging layer dispatches it to Knative Broker.

4. Knative Broker sends events to Knative Trigger which registered interest in receiving them.

> 5. Knative Trigger checks if the attributes of an incoming event match its specification. If they do, Knative Trigger sends the event to the Subscriber.

Author: 👍

## Knative Eventing Mesh component dependencies

This diagram shows how Eventing Mesh components work together.

![Eventing implementation](./assets/eventing-mesh-implementation.svg)

1. The user creates an [Application CR](https://kyma-project.io/docs/components/application-connector/#custom-resource-application) and binds it to a Namespace. 

2. The Application Operator watches the creation of the Application CR and creates an [HTTPSource CR](#custom-resource-http-source) which defines the source sending the events.

3. The Event Source Controller watches the creation of the HTTPSource CR and deploys these resources:

    * [HTTP Source Adapter](https://github.com/kyma-project/kyma/tree/master/components/event-sources/adapter/http) which is an HTTP server deployed inside the `kyma-integration` Namespace. This adapter acts as a gateway to the Channel, and is responsible for exposing an endpoint to which the Application sends the events. 

>     * [Knative Channel](https://knative.dev/docs/eventing/channels/) which defines the way messages are dispatched in the Namespace. Its underlying implementation is responsible for forwarding events to the Broker or additional Channels. Kyma uses NATS Streaming as its default Channel, however, you can change it to InMemoryChannel, Kafka, or Google PubSub. For details on configuring the default Channel, see [this](/components/event-bus/#details-configure-default-knative-channel) document. 

Reviewer 2: ```suggestion
    * [Knative Channel](https://knative.dev/docs/eventing/channels/) which defines the way messages are dispatched in the Namespace. Its underlying implementation is responsible for forwarding events to the Broker or additional Channels. Kyma uses NATS Streaming as its default Channel, but you can change it to InMemoryChannel, Kafka, or Google PubSub. For details on configuring the default Channel, see [this](/components/event-bus/#details-configure-default-knative-channel) document. 
```

Author: Done.

4. The Application Broker watches the creation of the Application CR and performs the following actions:

>     * Exposes the Events API which an external system provides, as a ServiceClass. Once the user provisions this ServiceClass in the Namespace, the Application Broker makes events available to use.

Reviewer 2: ```suggestion
    * Exposes the Events API of an external system as a ServiceClass. Once the user provisions this ServiceClass in the Namespace, the Application Broker makes events available to use.
```

Author: 👍

    * Deploys Knative Subscription and defines the Broker as the subscriber for the Channel to allow communication between them.

    * Adds the `knative-eventing-injection` label to the user's Namespace. As a result, the Namespace controller creates the [Broker](https://knative.dev/docs/eventing/broker-trigger/) which acts as an entry point for the events. 

   

> 

Reviewer 2: Could you remove these black lines?
